### PR TITLE
Support Unclip image variation

### PR DESCRIPTION
**Pull request by zhtmike**

### CODE REVIEW

**Review by hadipash — APPROVED** [review]

File: examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py [hunk 1]
```diff
@@ -352,9 +352,20 @@ def encode_first_stage(self, x):
     def get_first_stage_encoding(self, z):
         return self.scale_factor * z
 
-    def apply_model(self, x_noisy, t, c_concat=None, c_crossattn=None, return_ids=False, **kwargs):
-        x_recon = self.model(x_noisy, t, c_concat, c_crossattn, **kwargs)
-        return x_recon
+    def apply_model(self, x_noisy, t, cond, return_ids=False, **kwargs):
+        if isinstance(cond, dict):
+            # hybrid case, cond is expected to be a dict
+            pass
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py:358]: This change is aligned with change in https://github.com/Stability-AI/stablediffusion/blob/main/ldm/models/diffusion/ddpm.py#L849,  for easier migration of the official model in the future

File: examples/stable_diffusion_v2/README.md [hunk 2]
```diff
@@ -165,8 +167,46 @@ By setting empty prompt (`--prompt=""`), the masked part will be auto-filled to
 
 ### Text-guided Image-to-Image
 
-Coming soon
+This pipeline uses a finetuned version of Stable Diffusion 2.1, which can be used to create image variations (image-to-image). Here we provide two models `2.1-unclip-l` and `2.1-unclip-h`, which uses the pretrained Clip Image embedder and Open Clip Image embedder separately. User can use the `-v` argument to decide which model to use. The amount of image variation can be controled by the noise injected to the image embedding, which can be input by the `--noise_level` argument. (0 means no noise, 1000 full noise).
```
Comment (DongyaoZhu) [inline examples/stable_diffusion_v2/README.md:170]: CLIP, Users, controlled, ... please use grammarly / GPT to correct solecism.

File: examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DongyaoZhu) [reply examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py:358]: ddim.py, sampler.py, ... depend on this function, particularly a keyword arg.

File: examples/stable_diffusion_v2/ldm/modules/encoders/modules.py [hunk 3]
```diff
@@ -68,3 +75,171 @@ def encode(self, tokenized_text):
...
+    def encode(self, tokenized_text):
+        outputs = self.model(tokenized_text)
+        return outputs
+
+    def construct(self, c):
+        outputs = self.model(c)
+        return outputs
+
+
+class ClipImageEmbedder(nn.Cell):
```
Comment (DongyaoZhu) [inline examples/stable_diffusion_v2/ldm/modules/encoders/modules.py:119]: [minor] CLIP?

**Review by DongyaoZhu — CHANGES_REQUESTED** [review]
overall good, please check for dependencies and grammars

File: examples/stable_diffusion_v2/README.md (hunk 2 above)
Comment (zhtmike) [reply examples/stable_diffusion_v2/README.md]: Updated, thx

File: examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py:358]: Fixed dependencies

File: examples/stable_diffusion_v2/ldm/modules/encoders/modules.py (hunk 3 above)
Comment (zhtmike) [reply examples/stable_diffusion_v2/ldm/modules/encoders/modules.py]: Updated name

**Review by DongyaoZhu — APPROVED** [review]

File: examples/stable_diffusion_v2/README.md [hunk 4]
```diff
@@ -165,8 +167,46 @@ By setting empty prompt (`--prompt=""`), the masked part will be auto-filled to
...
 
+<div align="center">
+<img src="https://github.com/zhtmike/mindone/assets/8342575/251ea23e-92a5-4f52-a511-3bb11b36c92a" width="320" />
+  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
+<img src="https://github.com/zhtmike/mindone/assets/8342575/5c09e33e-1b4d-46c5-b4cf-659cdef26b77" width="320" />
+</div>
+<p align="center">
+<em> Image variation with prompt </em>
+  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
+<em> Image variation with prompt and noise </em>
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/README.md:208]: Please show the input image as well

File: examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py [hunk 5]
```diff
@@ -352,9 +352,20 @@ def encode_first_stage(self, x):
     def get_first_stage_encoding(self, z):
         return self.scale_factor * z
 
-    def apply_model(self, x_noisy, t, c_concat=None, c_crossattn=None, return_ids=False, **kwargs):
-        x_recon = self.model(x_noisy, t, c_concat, c_crossattn, **kwargs)
-        return x_recon
+    def apply_model(self, x_noisy, t, cond, return_ids=False, **kwargs):
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py:355]: I think we should preserve the explicit naming of `c_concat` and `c_crossattn`, referring to the discussion in https://github.com/mindspore-lab/mindone/pull/93#discussion_r1315332684

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py [hunk 6]
```diff
@@ -172,3 +173,90 @@ def construct(self, text):
...
+                )
+                for _ in range(layers)
+            ]
+        )
+
+    def construct(self, x, attn_mask):
+        return self.resblocks(x, attn_mask)
+
+
+class OpenClipTextEncoder(nn.Cell):
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:232]: how is it different from previous clip text encoder?

**Review by SamitHuang — COMMENTED** [review]
Nice work. Minor fix may needed.

File: examples/stable_diffusion_v2/README.md (hunk 4 above)
Comment (zhtmike) [reply examples/stable_diffusion_v2/README.md]: Update Image

File: examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py (hunk 5 above)
Comment (zhtmike) [reply examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py:355]: As discussed, the `cond` now can accept either a dictionary or a Tensor. A docstring is added to make it clearer.

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py (hunk 6 above)
Comment (zhtmike) [reply examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:232]: Basic there is no difference. But the writing style of `OpenClipTextEncoder` strictly follow what stablediffusion does (which relies on `CLIP` and `OpenCLIP` package), thus the migration of the checkpoint is easier in this case

**Review by SamitHuang — APPROVED** [review]
lgtm

**Pull request merged**